mpt_trie/src/trie_ops.rs

```diff
@@ -6,13 +6,41 @@ use std::{fmt::Display, mem::size_of};
 use enum_as_inner::EnumAsInner;
 use ethereum_types::{H256, U128, U256, U512};
 use log::trace;
+use thiserror::Error;
 
 use crate::{
     nibbles::{Nibble, Nibbles},
     partial_trie::{Node, PartialTrie, WrappedNode},
     utils::TrieNodeType,
 };
 
+/// Stores the result of trie operations. Returns a [TrieOpError] upon
+/// failure.
+pub type TrieOpResult<T> = Result<T, TrieOpError>;
+
+/// An error type for trie operation.
+#[derive(Debug, Error)]
+pub enum TrieOpError {
+    /// An error that occurs when a hash node is found during an insert
+    /// operation.
+    #[error("Found a `Hash` node during an insert in a `PartialTrie`! These should not be able to be traversed during an insert!")]
+    HashNodeInsertError(),
+
+    /// An error that occurs when a hash node is found during a delete
+    /// operation.
+    #[error("Attempted to delete a value that ended up inside a hash node")]
+    HashNodeDeleteError(),
+
+    /// An error that occurs when encontered an unexisting type of node during
+    /// an extension node collapse.
+    #[error("Extension managed to get an unexisting child node type! (child: {0})")]
+    HashNodeExtError(TrieNodeType),
+
+    /// Failed to insert a hash node into the trie.
+    #[error("Attempted to place a hash node on an existing node! (hash: {0})")]
+    ExistingHashNodeError(H256),
```

Comment:
I guess sort of adding on to my above comment, it's probably also worth having the key piece where these hash nodes were encountered at.